### PR TITLE
Update layout_map to use re.search instead of re.match.

### PR DESCRIPTION
**Pull request by qlzh727**
This allow user to skip the leading ".*' and make the rule more readable.

### CODE REVIEW

File: keras/distribution/distribution_lib.py [hunk 1]
```diff
@@ -327,10 +327,10 @@ class ModelParallel(Distribution):
     # will be split across 4 devices. Any other variable that doesn't
     # match any key in the layout map will be fully replicated.
     layout_map = LayoutMap(device_mesh)
-    layout_map['.*dense.*kernel'] = (None, 'model')
-    layout_map['.*dense.*bias'] = ('model',)
-    layout_map['.*conv2d.*kernel'] = (None, None, None, 'model')
-    layout_map['.*conv2d.*bias'] = ('model',)
+    layout_map['dense.*kernel'] = (None, 'model')
```
Comment (mattdangerw) [inline keras/distribution/distribution_lib.py:330]: what's the reason for `.*` in the middle instead of `/` again? the optimizer layout paths?

Comment (qlzh727) [reply keras/distribution/distribution_lib.py:330]: oh, that's for dense_1/dense_2.

Comment (mattdangerw) [reply keras/distribution/distribution_lib.py:330]: ah right thanks! so i guess that will be up to the user's practice. if they are setting the name themselves on all layers, that won't be necessary (I think?), if they are not, they should leave that in

**Review by mattdangerw — APPROVED** [review]
This looks good to me! I am trying to follow suit for the LoRA api (similar weight matching based on path required), so I can make the same change over there).

File: keras/distribution/distribution_lib.py (hunk 1 above)
Comment (qlzh727) [reply keras/distribution/distribution_lib.py:330]: Correct.

File: keras/distribution/distribution_lib.py [hunk 2]
```diff
@@ -443,22 +443,35 @@ def __getitem__(self, key):
         """Retrieves the corresponding layout by the string key.
 
         When there isn't an exact match, all the existing keys in the layout map
-        will be treated as a regex and map against the input key again. The
-        first match will be returned, based on the key insertion order. Returns
-        `None` if there isn't any match found.
+        will be treated as a regex and map against the input key again. When
+        there are multiple matches for the regex, an ValueError will be raised.
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply keras/distribution/distribution_lib.py]: Done.

File: keras/distribution/distribution_lib.py [hunk 3]
```diff
@@ -443,22 +443,35 @@ def __getitem__(self, key):
...
             Corresponding layout based on the query.
+
+        Raises:
+            ValueError when multiple keys are matched if the keys are treated
+            as regex.
         """
         if key in self._layout_map:
             return self._layout_map[key]
 
+        matching_key = []
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply keras/distribution/distribution_lib.py]: Done.

File: keras/distribution/distribution_lib.py [hunk 4]
```diff
@@ -443,22 +443,35 @@ def __getitem__(self, key):
...
+        there are multiple matches for the regex, an ValueError will be raised.
+        Returns `None` if there isn't any match found.
 
         Args:
             key: String key to query a layout.
 
         Returns:
             Corresponding layout based on the query.
+
+        Raises:
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply keras/distribution/distribution_lib.py]: Done.

File: keras/distribution/distribution_lib.py [hunk 5]
```diff
@@ -443,22 +443,35 @@ def __getitem__(self, key):
...
         for k in self._layout_map:
-            if re.match(k, key):
-                return self._layout_map[k]
+            if re.search(k, key):
+                matching_key.append(k)
+        if len(matching_key) > 1:
+            raise ValueError(
+                f"The input {key} has matched to multiple layout "
+                f"rule: {matching_key}. Please make sure the "
+                "key only match to one rule."
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply keras/distribution/distribution_lib.py]: Thanks. Done.

**Review by fchollet — APPROVED** [review]
LGTM

**Pull request merged**